WebApp Updates

evd123 opened the pull request:
-Finished HTML and CSS components
-Still need to work on JS

evd123 committed:
All commits from start.

Add fact button to main page.

Add columns with HTML and CSS; change colors and add gradient

Update content, add card style, and add emojis

Update color, add links, fix columns

Add images

Remove files in target/ from commits (hopefully)

More formatting fixes

Reformat, add changes

lcatherine commented:
Could you squash the commits and add a screenshot of what the webapp looks like?

evd123 commented:
I tried following: https://help.github.com/en/github/using-git/using-git-rebase-on-the-command-line#pushing-rebased-code-to-github
but ran into some issues... 
I tried: git rebase -i <commit #> with a couple of commits. Then, the link (above) said to save and close the editor and Git would reopen the editor. I'm not sure if Cloud Shell works exactly the same way as the terminal, because it didn't seem to reopen another editor. Then, I aborted the rebase to try again, and now get:

noop

Rebase ff40e76..ff40e76 onto ff40e76 (1 command)

Commands:
p, pick = use commit
r, reword = use commit, but edit the commit message
e, edit = use commit, but stop for amending
s, squash = use commit, but meld into previous commit
f, fixup = like "squash", but discard this commit's log message
x, exec = run command (the rest of the line) using shell
d, drop = remove commit

These lines can be re-ordered; they are executed from top to bottom.

If you remove a line here THAT COMMIT WILL BE LOST.

However, if you remove everything, the rebase will be aborted.

I'm assuming I rebased and aborted too many times. 
Should I try this: https://docs.gitlab.com/ee/user/project/merge_requests/squash_and_merge.html?

I added a screenshot of the webapp to Settings -> Social Preview

lcatherine commented:
Actually that's correct. 

git rebase -i 

Will automatically show you the list of all the commits since head that have yet to be pushed.

(-i means interactive mode)

You can then change the keywords in front of them to squash everything except for the most recent commit, then change the commit message to something that encapsulates what is now in the new commit.

https://git-scm.com/docs/git-rebase#_interactive_mode

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -27,6 +29,41 @@
+  /**
+   * Converts a ServerStats instance into a JSON string using manual String concatentation.
+   */
+  private String convertToJson(ArrayList<String> messages) {
```
This produces the following output:

```{How's it going? My name is Emma. I'm a junior at Rice.}```

It's not quite JSON. Is this intended?

lyze reviewed (commented):
Can we name the images more descriptively?

lyze commented on portfolio/src/main/webapp/index.html:
```diff
@@ -151,10 +151,29 @@ <h3>Contact</h3>
+            <input type="checkbox" name="yes-response" value="true"> Check if you want a response!
+
+            <p>Email:</p>
+            <textarea name="email-input">Email address</textarea>
```
`<input type="email" placeholder="me@example.com" ...`

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -41,6 +41,30 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+    String email = getEmail(request);
+    response.getWriter().println("Thanks for reaching out!");
+
+    if (yes && email == "Email address") {
```
it's a little odd to check against this magic string. I would have expected to check against the empty string. Perhaps the input change below may make it possible to change this check.

evd123 commented:
![image](https://user-images.githubusercontent.com/46075902/83764482-fca85300-a647-11ea-9c52-c3c555be0421.png)

When trying to squash commits, I am now getting this image (I'm using git rebase -i Emma_Development).

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,102 @@
+    }
+    // Send the JSON as the response
+    response.setContentType("application/json");
+    String json = new Gson().toJson(allMessages);
```
Gson gson = new Gson();
gson.toJson(allMessages, gson.newJsonWriter(response.writer()));

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,102 @@
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+    }
```
nit: fix indentation on this line

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,102 @@
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(messageEntity);
+
+    // original way of adding messages to list:
```
remove leftover code?

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,102 @@
+      if (emailString == null) {
+          System.err.println("It looks like you're interested in a response! Please enter your email address so that I can get in touch.");
+          return "false";
+        }
```
indentation appears off

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DeleteMessageServlet.java:
```diff
@@ -0,0 +1,39 @@
+    Key messageEntityKey = KeyFactory.createKey("Message", id);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.delete(messageEntityKey);
+  }
```
also here

let's run this through the code formatter

lyze commented on portfolio/target/maven-archiver/pom.properties:
```diff
@@ -0,0 +1,5 @@
+#Generated by Maven
```
all of the files in `portfolio/target/...` should be excluded from commits, since they're generated as part of the build. let's add target/ to gitignore.

evd123 replied:
It looks like target/ is already in gitignore... I think I accidentally moved a file out of target and then had to move it back in which is why this happened

lyze reviewed (changes requested):
Let's make sure to exclude the target files before merging, and also squash when we merge it in.

lcatherine commented on portfolio/pom.xml:
```diff
@@ -23,6 +23,16 @@
+      <version>2.8.6</version>
+    </dependency>
+    <dependency>
+    <groupId>com.google.appengine</groupId>
+    <artifactId>appengine-api-1.0-sdk</artifactId>
+    <version>1.9.59</version>
```
indentation

lcatherine commented on portfolio/pom.xml:
```diff
@@ -35,7 +45,7 @@
         <version>2.2.0</version>
         <configuration>
           <!-- TODO: set project ID. -->
```
You can remove this since project id is set

lcatherine commented on portfolio/src/main/java/com/google/sps/data/Message.java:
```diff
@@ -0,0 +1,18 @@
+    private final String email;
+    private final long timestamp;
+
+    /** Add the new message to the list of all messages */
```
Is this comment in the wrong place? The associated code is for Message constructor

lcatherine commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,94 @@
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        List < Message > allMessages = new ArrayList < > ();
```
`List < Message >` extra spacing

lcatherine replied:
`ArrayList < > ()` extra spacing

lcatherine commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,19 +14,94 @@
+    /*
+    Function to check if email is submitted if the user wants a response. Don't think this function actually does anything.
+    */
+    private String getYes(HttpServletRequest request) {
```
you can have this return a bool to make the check easier

lcatherine reviewed (dismissed):
Some minor comments

lyze commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -56,24 +56,23 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         // Send the JSON as the response
         response.setContentType("application/json");
         Gson gson = new Gson();
-        String json = gson.toJson(allMessages, gson.newJsonWriter(response.writer()));
```
```gson.toJson(allMessages, gson.newJsonWriter(response.writer()));```

is all we need; we don't need `response.getWriter().println(...)` since passing in the responsewriter means gson will write the output there. This also is a bit more efficient so that we don't create an intermediate string object.

lyze commented on portfolio/src/main/webapp/script.js:
```diff
@@ -11,6 +11,10 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
+function initialize() {
```
let's move development of the maps API into a new PR branch to keep things clean. let's look to finish up the original PR and get that merged in.

evd123 commented:
I tried another formatter that looked like it followed the public style guide more closely and fixed the gson/json syntax and I think I removed the target/ file from the commits!
